### PR TITLE
【PaddlePaddle Hackathon】在Paddle2ONNX 新增11个 Paddle 2.0 API 支持

### PR DESCRIPTION
**Pull request by AndPuQing**
11个 API 具体为：

paddle.nn.functional.interpolate(x, size=None, scale_factor=None, mode='nearest', align_corners=False, align_mode=0, data_format='NCHW', name=None)

paddle.nn.functional.softshrink(x, threshold=0.5, name=None)

paddle.tan(x, name=None)

paddle.nn.Tanhshrink(name=None)

paddle.nn.functional.thresholded_relu(x, threshold=1.0, name=None)

paddle.unique(x, return_index=False, return_inverse=False, return_counts=False, axis=None, dtype='int64', name=None)

paddle.where(condition, x, y, name=None)

paddle.scatter(x, index, updates, overwrite=True, name=None)

paddle.scatter_nd_add(x, index, updates, name=None)

paddle.meshgrid(*args, **kargs)

paddle.masked_select(x, mask, name=None)

### CODE REVIEW

**Comment by AndPuQing** [comment]
## interpolate

1. AREA 面积插值 并未修改，因调用到了 `adaptive_avg_pool`
2. 对opset 11添加了对`data_format`设置的支持
3. 修改`NEAREST`转换设置中`coordinate_transformation_mode`  为 `asymmetric`，使其精度对齐
4. 添加对于`linear`与`bicubic`的支持

## scatter

scatter PaddlePaddle中有`overwrite`属性，可以对当有相同索引情况进行设置。当`overwrite = True` 时与`onnx`相同，当`overwrite = False`时opset 15并无相似设置进行对齐，之后`onnx`对其进行了支持，可以见https://github.com/onnx/onnx/pull/3484 提交。但是`onnxruntime`还未支持到该版本（即opset 16）。

## scatter_nd_add

类似于上面`scatter`，当有重复索引时精度不会对齐

**Comment by jiangjiajun** [comment]
> ## interpolate
> 1. AREA 面积插值 并未修改，因调用到了 `adaptive_avg_pool`
> 2. 对opset 11添加了对`data_format`设置的支持
> 3. 修改`NEAREST`转换设置中`coordinate_transformation_mode`  为 `asymmetric`，使其精度对齐
> 4. 添加对于`linear`与`bicubic`的支持
> 
> ## scatter
> scatter PaddlePaddle中有`overwrite`属性，可以对当有相同索引情况进行设置。当`overwrite = True` 时与`onnx`相同，当`overwrite = False`时opset 15并无相似设置进行对齐，之后`onnx`对其进行了支持，可以见[onnx/onnx#3484](https://github.com/onnx/onnx/pull/3484) 提交。但是`onnxruntime`还未支持到该版本（即opset 16）。
> 
> ## scatter_nd_add
> 类似于上面`scatter`，当有重复索引时精度不会对齐

scatter的实现参考https://github.com/apache/tvm/blob/main/python/tvm/relay/frontend/paddlepaddle.py#L1107

**Comment by AndPuQing** [comment]
> > ## interpolate
> > 
> > 1. AREA 面积插值 并未修改，因调用到了 `adaptive_avg_pool`
> > 2. 对opset 11添加了对`data_format`设置的支持
> > 3. 修改`NEAREST`转换设置中`coordinate_transformation_mode`  为 `asymmetric`，使其精度对齐
> > 4. 添加对于`linear`与`bicubic`的支持
> > 
> > ## scatter
> > scatter PaddlePaddle中有`overwrite`属性，可以对当有相同索引情况进行设置。当`overwrite = True` 时与`onnx`相同，当`overwrite = False`时opset 15并无相似设置进行对齐，之后`onnx`对其进行了支持，可以见[onnx/onnx#3484](https://github.com/onnx/onnx/pull/3484) 提交。但是`onnxruntime`还未支持到该版本（即opset 16）。
> > ## scatter_nd_add
> > 类似于上面`scatter`，当有重复索引时精度不会对齐
> 
> scatter的实现参考https://github.com/apache/tvm/blob/main/python/tvm/relay/frontend/paddlepaddle.py#L1107

`overwrite = False`似乎用到了下面这个`api`

https://github.com/apache/tvm/blob/86781e99371663f7c3315f8d98aba273dddb4422/python/tvm/relay/frontend/paddlepaddle.py#L1123

其中的https://github.com/apache/tvm/blob/86781e99371663f7c3315f8d98aba273dddb4422/python/tvm/topi/cuda/scatter.py#L716 这句似乎是将每个index抽出计算，然后合并？？？

**Comment by jiangjiajun** [comment]
不是指看TVM的OP TOPI实现，而是说在Paddle OP->TVM Relay IR时，在overwrite不为True的情况下，上面的代码使用了OP组合来完成

现在在Paddle2ONNX中，是否也可以这样做呢

File: paddle2onnx/op_mapper/op_mapper.py [hunk 1]
```diff
@@ -91,6 +91,7 @@ def __call__(self, cls):
                         OpMapper.OPSETS[op] = {}
                     opset_dict = OpMapper.OPSETS[op]
                     opset_dict[version] = (v, self.kwargs)
+        return cls
```
Comment (AndPuQing) [inline paddle2onnx/op_mapper/op_mapper.py]: 返回修饰的类对象

**Comment by AndPuQing** [comment]
> 不是指看TVM的OP TOPI实现，而是说在Paddle OP->TVM Relay IR时，在overwrite不为True的情况下，上面的代码使用了OP组合来完成
> 
> 现在在Paddle2ONNX中，是否也可以这样做呢

您好，最新提交中更新了对OP组合的支持，但是实现方式比较粗糙，不知是否有更好的方式实现。

**Comment by yeliang2258** [comment]
test_nn_functional_interpolate_linear_scale_factor_float这个单测有diff，请核对一下是什么原因

**Comment by AndPuQing** [comment]
> test_nn_functional_interpolate_linear_scale_factor_float这个单测有diff，请核对一下是什么原因

你好，是否可以具体看看报错位置，尚不清楚报错原因

**Comment by yeliang2258** [comment]
> > test_nn_functional_interpolate_linear_scale_factor_float这个单测有diff，请核对一下是什么原因
> 
> 你好，是否可以具体看看报错位置，尚不清楚报错原因

image
就是转换后的onnxruntime得到的结果和paddle的结果不一致，有diff

**Comment by yeliang2258** [comment]
> > test_nn_functional_interpolate_linear_scale_factor_float这个单测有diff，请核对一下是什么原因
> 
> 你好，是否可以具体看看报错位置，尚不清楚报错原因

你线下测试的时候是否复现了问题？单测都可以跑过吗？ONNX和ONNXRuntime版本是？

**Comment by AndPuQing** [comment]
> > > test_nn_functional_interpolate_linear_scale_factor_float这个单测有diff，请核对一下是什么原因
> > 
> > 
> > 你好，是否可以具体看看报错位置，尚不清楚报错原因
> 
> 你线下测试的时候是否复现了问题？单测都可以跑过吗？ONNX和ONNXRuntime版本是？

您好，我在测试时，误差在2E-6左右，`ONNX==1.10.1`，`onnxruntime==1.9.0`，现在推断是onnx对于一次线性插值是需要`2DTensor`输入的，但是paddle输入是`3DTensor`，所以是ONNX对其是二/三次插值，所以精度没有对齐

**Comment by yeliang2258** [comment]
image
CI这里还是有错，请再核查一下

**Comment by AndPuQing** [comment]
奇怪，`tile`算子在提交中并未修改，以下是我单测结果
image

**Comment by yeliang2258** [comment]
> 算子在提交中并未修改，以下是我单测结果

那有可能是环境问题，你跑一遍所有的单侧，看能否全都通过。

**Comment by AndPuQing** [comment]
> > 算子在提交中并未修改，以下是我单测结果
> 
> 那有可能是环境问题，你跑一遍所有的单侧，看能否全都通过。

您好，确认是版本的问题，我开发一直是`ONNX == 1.10.0`，所以在写单测时，支持版本会超过`ONNX == 1.9.0`，造成错误

**Comment by yeliang2258** [comment]
好的，如果你这边所有单测都可以跑过，那明天我们按照ONNX==1.10.1，onnxruntime==1.9.0来搭建环境再跑一下试试。



---Original---
From: ***@***.***&gt;
Date: Wed, Nov 17, 2021 21:58 PM
To: ***@***.***&gt;;
Cc: ***@***.***&gt;;"Review ***@***.***&gt;;
Subject: Re: [PaddlePaddle/Paddle2ONNX] Add ten operator support and unittesting (PR #392)




   
算子在提交中并未修改，以下是我单测结果
  
那有可能是环境问题，你跑一遍所有的单侧，看能否全都通过。
  
您好，确认是版本的问题，我开发一直是ONNX == 1.10.0，所以在写单测时，支持版本会超过ONNX == 1.9.0，造成错误
 
—
You are receiving this because your review was requested.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by yeliang2258** [comment]
你好，请git commit –allow-empty再git push origin xxxx提交一个空commit，重新跑一次CI

**Comment by AndPuQing** [comment]
你好，我测试环境是
image

跑完所有单测
image

不知现在报错是什么，我会尽快尝试解决

**Comment by Zeref996** [comment]
你好，现在CI的api单测是通过了，但是某些模型转换出现了以下报错：
56ae088de2c8d41be2e2342e6677e58d
导致以下模型opt版本9和10转换失败（opt版本11模型都passed）：
image
全部报错信息：
image

**Comment by Zeref996** [comment]
以下命令可下载出现问题的3个infer模型：
wget https://paddle-qa.bj.bcebos.com/Paddle2ONNX/user/DeepLabv3.tar.gz
wget https://paddle-qa.bj.bcebos.com/Paddle2ONNX/user/HRNet_Seg.tar.gz
wget https://paddle-qa.bj.bcebos.com/Paddle2ONNX/user/UNet.tar.gz

使用环境：
image
paddlepaddle==develop
paddle2onnx使用setup编译安装

执行以下脚本可将infer模型转换为onnx模型：
paddle2onnx --model_dir DeepLabv3 \
          --model_filename DeepLabv3/__model__ \
          --params_filename DeepLabv3/__params__ \
          -s DeepLabv3/DeepLabv3_opt9.onnx  \
          --opset_version 9

paddle2onnx --model_dir DeepLabv3 \
          --model_filename DeepLabv3/__model__ \
          --params_filename DeepLabv3/__params__ \
          -s DeepLabv3/DeepLabv3_opt10.onnx  \
          --opset_version 10

paddle2onnx --model_dir DeepLabv3 \
          --model_filename DeepLabv3/__model__ \
          --params_filename DeepLabv3/__params__ \
          -s DeepLabv3/DeepLabv3_opt11.onnx  \
          --opset_version 11

转换opt 9，opt 10版本时可复现报错

**Comment by yeliang2258** [comment]
开发者你好，请问现在你那边解决这个单测有什么问题我们可以提供帮助的吗？

**Comment by yeliang2258** [comment]
现在单测显示提交的代码有几个模型转换失败，你可以将上面的三个模型wget下来，然后测试是否可以转换，以及转换之后ONNXRuntime和Paddle的diff小于1e-5

**Comment by AndPuQing** [comment]
> 现在单测显示提交的代码有几个模型转换失败，你可以将上面的三个模型wget下来，然后测试是否可以转换，以及转换之后ONNXRuntime和Paddle的diff小于1e-5

Sorry for the late reply, the error was caused by a problem passing arguments into the `compute_output_shape` function, which has been fixed

File: paddle2onnx/op_mapper/math.py [hunk 2]
```diff
@@ -146,6 +146,16 @@ def opset_7(cls, graph, node, **kw):
             'Atan', inputs=node.input('X'), outputs=node.output('Out'))
 
 
+@op_mapper('tan')
```
Comment (jiangjiajun) [inline paddle2onnx/op_mapper/math.py:149]: https://github.com/PaddlePaddle/Paddle2ONNX/blob/develop/paddle2onnx/op_mapper/activation.py#L24 tanh移到此处实现

File: paddle2onnx/op_mapper/op_mapper.py (hunk 1 above)
Comment (jiangjiajun) [reply paddle2onnx/op_mapper/op_mapper.py]: 这里返回的作用是？

**Review by jiangjiajun — CHANGES_REQUESTED** [review]

File: paddle2onnx/op_mapper/op_mapper.py (hunk 1 above)
Comment (AndPuQing) [reply paddle2onnx/op_mapper/op_mapper.py]: 你好，这里返回是因为上方讨论是否可以使用OP组合的方式实现，返回后在OP组合实现才可以拿到方法

Comment (jiangjiajun) [reply paddle2onnx/op_mapper/op_mapper.py]: 看了下实现，你在里面调用了另外一个OP的转换代码，我先前讨论中提到的组合是指用ONNX的OP来组合实现，而不是指在scatter的转换时，再调用scatterndadd的转换代码。 
当前的修改会使得框架转换逻辑复杂化，并且OP的耦合也可能会提升未来维护成本。

建议将这里的返回删除，然后scatter的转换改为直接调用onnx的op来组合

**Comment by jiangjiajun** [comment]
为了推动v1.0的快速发版，此PR先合入，剩余问题将会内部再解决。
